ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Query.hs

```diff
@@ -44,54 +53,106 @@ import           Ouroboros.Consensus.Util.DepPair
+--
+-- Additions to the set of queries is versioned by 'QueryVersion'
+data Query blk result where
+  -- | This constructor is supported by all @QueryVersion@s. The @BlockQuery@
   -- argument is versioned by the @BlockNodeToClientVersion blk@.
+  BlockQuery :: BlockQuery blk result -> Query blk result
+
+  -- | Get the 'SystemStart' time.
+  --
+  -- Supported by 'QueryVersion' >= 'QueryVersion1'.
+  GetSystemStart :: Query blk SystemStart
 
 instance (ShowProxy (BlockQuery blk)) => ShowProxy (Query blk) where
   showProxy (Proxy :: Proxy (Query blk)) = "Query (" ++ showProxy (Proxy @(BlockQuery blk)) ++ ")"
 
 instance (ShowQuery (BlockQuery blk)) => ShowQuery (Query blk) where
   showResult (BlockQuery blockQuery) = showResult blockQuery
+  showResult GetSystemStart          = show
 
 instance Eq (SomeSecond BlockQuery blk) => Eq (SomeSecond Query blk) where
   SomeSecond (BlockQuery blockQueryA) == SomeSecond (BlockQuery blockQueryB)
     = SomeSecond blockQueryA == SomeSecond blockQueryB
+  SomeSecond (BlockQuery _) == _ = False
+
+  SomeSecond GetSystemStart == SomeSecond GetSystemStart = True
+  SomeSecond GetSystemStart == _                         = False
 
 instance Show (SomeSecond BlockQuery blk) => Show (SomeSecond Query blk) where
   show (SomeSecond (BlockQuery blockQueryA)) = "Query " ++ show (SomeSecond blockQueryA)
+  show (SomeSecond GetSystemStart) = "Query GetSystemStart"
+
+
+-- | Exception thrown in the encoders
+data QueryEncoderException blk =
+    -- | A query was submitted that is not supported by the given 'QueryVersion'
+    QueryEncoderUnsupportedQuery
+         (SomeSecond Query blk)
+         QueryVersion
+
+deriving instance Show (SomeSecond BlockQuery blk) => Show (QueryEncoderException blk)
+instance (Typeable blk, Show (SomeSecond BlockQuery blk)) => Exception (QueryEncoderException blk)
 
 queryEncodeNodeToClient ::
-     forall blk. SerialiseNodeToClient blk (SomeSecond BlockQuery blk)
+     forall blk.
+     Typeable blk
+  => Show (SomeSecond BlockQuery blk)
+  => SerialiseNodeToClient blk (SomeSecond BlockQuery blk)
   => CodecConfig blk
   -> QueryVersion
   -> BlockNodeToClientVersion blk
   -> SomeSecond Query blk
   -> Encoding
 queryEncodeNodeToClient codecConfig queryVersion blockVersion (SomeSecond query)
   = case queryVersion of
-    TopLevelQueryDisabled -> encodeBlockQuery
-      (case query of
-        BlockQuery blockQuery -> blockQuery
-      )
+
+    -- In "version 0" we only support BlockQuery and add no extra wrapping so
+    -- that it's backwards compatible with when there were no top level queries.
+    TopLevelQueryDisabled ->
+      case query of
+        BlockQuery blockQuery ->
+          encodeBlockQuery blockQuery
+        _ ->
+          throw $ QueryEncoderUnsupportedQuery (SomeSecond query) queryVersion
+
+    -- From version 1 onwards, we use normal constructor tags
+    QueryVersion1 ->
+      case query of
+        BlockQuery blockQuery ->
+            encodeListLen 2
```

Comment:
I've seen this encoding with the list length in a few places, but I don't understand the motivation. Perhaps it's just for an extra internal check, but that seems like an unconvincing reason to me.

Reply:
Valid CBOR requires us to use structures (lists or maps) for multiple values.